policy/caa.go

```diff
@@ -0,0 +1,282 @@
+}
+
+// DNS utility methods
+
+// DNSKEY record for "." so we don't have to query it every time we want to
+// build a chain of trust.
+const rootDNSKey = `			75417	IN	DNSKEY	257 3 8 AwEAAagAIKlVZrpC6Ia7gEzahOR+9W29euxhJhVVLOyQbSEW0O8gcCjF FVQUTf6v58fLjwBd0YI0EzrAcQqBGCzh/RStIoO8g0NfnfL2MTJRkxoX bfDaUeVPQuYEhg37NZWAJQ9VnMVDxP/VHL496M/QZxkjf5/Efucp2gaD X6RS6CXpoY68LsvPVjR0ZSwzz1apAzvN9dlzEheX7ICJBBtuA6G3LQpz W5hOA2hzCTMjJPJ8LbqF6dsV6DoBQzgul0sGIcGOYl7OyQdXfZ57relS Qageu+ipAdTTJ25AsRTAoub8ONGcLmqrAmRLKBP1dfwhYB4N7knNnulq QxA+Uk1ihz0=`
+
+func getNsAndRoot(pubU *unbound.Unbound, domain string) (string, string, error) {
+	// get NS records for domain
+	r, err := pubU.Resolve(dns.Fqdn(domain), dns.TypeNS, dns.ClassINET)
+	if err != nil {
+		return "", "", fmt.Errorf("Resolving '%s' failed: %s", domain, err)
+	}
+
+	nsName := ""
+	rootDomain := ""
+	if r.HaveData {
+		for _, record := range r.Rr {
+			if record.Header().Rrtype == dns.TypeNS {
+				// just grab the first NS domain (this breaks things when the first
+				// ns server returned is broken... ie offline/non-responsive on :53)
+				nsName = record.(*dns.NS).Ns
+				rootDomain = record.(*dns.NS).Hdr.Name
+				break
+			}
+		}
+	} else {
+		// if no NS records are returned its probably because domain is a CNAME,
+		// grab the authoritative NS server from the SOA record in the
+		// authority section of the answer.
+		if len(r.AnswerPacket.Ns) > 0 {
+			for _, record := range r.AnswerPacket.Ns {
+				if record.Header().Rrtype == dns.TypeSOA {
+					nsName = record.(*dns.SOA).Ns
+					rootDomain = record.(*dns.SOA).Hdr.Name
+					break
+				}
+			}
+			// silly ns in SOA...
+			if nsName == "." {
+				for _, record := range r.AnswerPacket.Ns {
+					if record.Header().Rrtype == dns.TypeSOA {
+						r, err := pubU.Resolve(record.(*dns.SOA).Hdr.Name, dns.TypeNS, dns.ClassINET)
+						if err != nil {
+							return "", "", fmt.Errorf("Resolving '%s' failed: %s", record.(*dns.SOA).Hdr.Name, err)
+						}
+						if r.HaveData {
+							for _, nsRecord := range r.Rr {
+								if nsRecord.Header().Rrtype == dns.TypeNS {
+									// just grab the first NS domain
+									nsName = nsRecord.(*dns.NS).Ns
+									rootDomain = nsRecord.(*dns.NS).Hdr.Name
+									break
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	if nsName == "" || rootDomain == "" {
+		return "", "", fmt.Errorf("Couldn't retrieve authoritative nameserver or zone root for '%s'", domain)
+	}
+	rootDomain = strings.TrimRight(rootDomain, ".")
+
+	nsIPs, err := pubU.LookupIP(nsName)
+	if err != nil {
+		return "", "", fmt.Errorf("Resolving '%s' failed: %s", nsName, err)
+	}
+	if len(nsIPs) > 0 {
+		// return first IP for NS
+		return nsIPs[0].String(), rootDomain, nil
+	}
+	return "", "", fmt.Errorf("Address lookup did not return any IPs for '%s'", nsName)
+}
+
+func getDNSKeys(pubU *unbound.Unbound, authU *unbound.Unbound, domain string, rootDomain string) (bool, error) {
```

Comment:
I don't understand the boundary between this code and libunbound. It seems like libunbound does some amount of its own DNSSEC validation, as shown in bogus flag in Result, used in getCAA. Why do we also fetch DNSKEYS and manually validate them here?

That said, per my previous comment about libunbound, if having some of this logic in Go allows us to switch out a memory-safe library instead of libunbound, then this is probably a good idea.